static/selector.js

```diff
@@ -0,0 +1,81 @@
+/* eslint-disable quotes */
+/**
+ * Creates the selectors for the chart and updates the chart based on the
+ * selected values.
+ * @param {string} url The url for the application.
+ * @param {string} chartId The id of the chart being drawn.
+ * @param {*} colorScale A d3 colorscale used for the line colors.
+ */
+const selectors = async (url, chartId, colorScale) => {
+  const modes = ["Default", "K-means", "DBSCAN"];
+  const similarity = ["Correlation", "Proximity"];
+  let clusters = ["All"];
+  createSelector("mode", modes, updateChart);
+  createSelector("similarity", similarity, updateChart);
+  createSelector("cluster", clusters, updateCluster);
+
+  /**
+   * Updates the chart according to the values of the selectors.
+   */
+  async function updateChart() {
+    const currentMode = d3.select("select#modeSelector").property("value");
+    const currentSimilarity = d3.select("select#similaritySelector")
+        .property("value");
+    createSelector("cluster", ["All"], updateCluster);
+    updateCluster();
+
+    if (currentMode == "Default") {
+      d3.selectAll(".timeSeries")
+          .attr("stroke", (d) => colorScale(d))
+          .attr("opacity", 1);
+    } else {
+      const query = currentMode + "/" + currentSimilarity + "/" + chartId;
+      const response = await fetch(url + "clustering/" + query);
+      if (response.status >= 200 && response.status <= 299) {
+        const data = await response.json();
+        data.forEach((elt, index) => {
+          d3.selectAll("#id" + index)
+              .attr("stroke", colorScale(elt))
+              .attr("class", "timeSeries " + "cluster-All " +
+                            "cluster-" + elt);
+        });
+        clusters = ["All"].concat(Array.from(new Set(data)))
+            .sort((a, b) => a-b);
+        createSelector("cluster", clusters, updateCluster);
+      } else {
+        showError(response.status);
+      }
+    }
+  }
+
+  /**
+   * Update the selector element called name with options.
+   * @param {string} name The name of the selector element.
+   * @param {Array} options The options of the selector element.
+   * @param {string} updateFunction The function for updating the selector.
+   */
+  function createSelector(name, options, updateFunction) {
+    const selector = d3.select("select#"+ name + "Selector");
+    selector.selectAll("option")
+        .data(options)
+        .join("option")
+        .attr("value", (d) => d)
+        .text((d) => d);
+
+    selector.on("change", updateFunction);
```

Comment:
Do you know if selectAll clears on change callbacks? If not, calling createSelector multiple times could add redundant change handlers.

Reply:
selectAll does not clear on change call backs. It might make sense to move setting the handler outside of createSelector since this should only be done once.